### PR TITLE
[REEF-1665] Better default value for Java runtime path

### PR DESCRIPTION
**Pull request by motus**
Summary of changes:
  * Fix the default path to java binary use `{{JAVA_HOME}}` notation for all OS
  * Implement `{{...}}` variable expansion for REEF local runtime
  * Write unit tests for various cases of the variable expansion
  * Minor code refactoring for better readability

This fix allows to run `runtests.ps1 -Yarn ...` again.

JIRA: [REEF-1665](https://issues.apache.org/jira/browse/REEF-1665)

### CODE REVIEW

**Comment by zerg-junior** [comment]
@motus I believe this patch also affects [the prerequisites part of the Java build instructions](https://github.com/apache/reef/blob/master/lang/java/BUILD.md#prerequisites) and any other similar documentation we may have.

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/uploader/JobFolder.java [hunk 1]
```diff
@@ -57,53 +57,57 @@
...
-    } catch (final IOException e) {
-      LOG.log(Level.SEVERE, "Unable to upload {0} to {1}", new Object[]{source, destination});
-      throw e;
+    } catch (final IOException ex) {
+      LOG.log(Level.SEVERE, "Unable to upload " + source + " to " + destination, ex);
+      throw ex;
     }
-    LOG.log(Level.FINE, "Uploaded {0} to {1}", new Object[]{source, destination});
+
+    LOG.log(Level.FINE, "Uploaded {0} to {1}", new Object[] {source, destination});
```
Comment (zerg-junior) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/uploader/JobFolder.java]: For the sake of consistency, can you please alter this log statement to use the syntax without `Object[]` ?

File: lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/launch/JavaLaunchCommandBuilder.java [hunk 2]
```diff
@@ -37,19 +37,22 @@
  * Build the launch command for Java REEF processes.
  */
 public final class JavaLaunchCommandBuilder implements LaunchCommandBuilder {
+
   private static final Logger LOG = Logger.getLogger(JavaLaunchCommandBuilder.class.getName());
 
-  private static final String DEFAULT_JAVA_PATH = System.getenv("JAVA_HOME") + "/bin/" + "java";
+  private static final String DEFAULT_JAVA_PATH = "java"; // Assume java binary is in the PATH
```
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/launch/JavaLaunchCommandBuilder.java]: This is tricky. What if `java` is not on the path or if the right `java` is not on the path? Can we check how Hadoop  and Spark handle this and follow their lead? That way, we confuse users the least.

Comment (motus) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/launch/JavaLaunchCommandBuilder.java]: good point @markusweimer - let me see how Spark is handling that

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/uploader/JobFolder.java (hunk 1 above)
Comment (motus) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/uploader/JobFolder.java]: @zerg-junior I think we should stick to the `Object[] {}` wherever possible. This way we do not construct the log string if the message's priority is below the current log level. I use string concatenation in the previous log statement to make it print the exception stack trace.

File: lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/launch/JavaLaunchCommandBuilder.java (hunk 2 above)
Comment (motus) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/launch/JavaLaunchCommandBuilder.java]: I've just run our HelloREEF application on CISL YARN cluster with `{{JAVA_HOME}}/bin/java` and then with `%JAVA_HOME%/bin/java` - and in **both** cases it works! I bet Spark just uses one of those notations - will give you a definite answer in a minute

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/uploader/JobFolder.java (hunk 1 above)
Comment (zerg-junior) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/uploader/JobFolder.java]: A, I didn't know that. than it is surely fine.

File: lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/launch/JavaLaunchCommandBuilder.java (hunk 2 above)
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/launch/JavaLaunchCommandBuilder.java]: Yes, we should rely on established mechanisms for this. REEF should run if Hadoop and Spark run :)

Comment (motus) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/launch/JavaLaunchCommandBuilder.java]: @markusweimer Yep. I think we should use `{{JAVA_PATH}}` notation in REEF job submission and implement that variable expansion in the local runtime, too. I will submit a new update for it shortly.

**Comment by motus** [comment]
OK folks, here's a new iteration of the change. We now use `{{...}}`  syntax when building path for the remote java binary on the client. When the job is submitted, Hadoop performs that expansion for us; in local mode, I've implemented that expansion in the `RunnableProcess` class in REEF. @markusweimer could you please take a look? Thanks!

File: lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/launch/JavaLaunchCommandBuilder.java [hunk 3]
```diff
@@ -37,19 +37,22 @@
...
  */
 public final class JavaLaunchCommandBuilder implements LaunchCommandBuilder {
+
   private static final Logger LOG = Logger.getLogger(JavaLaunchCommandBuilder.class.getName());
 
-  private static final String DEFAULT_JAVA_PATH = System.getenv("JAVA_HOME") + "/bin/" + "java";
+  private static final String DEFAULT_JAVA_PATH = "{{JAVA_HOME}}/bin/java";
   private static final String[] DEFAULT_OPTIONS = {"-XX:PermSize=128m", "-XX:MaxPermSize=128m"};
+
+  private final Map<String, JVMOption> options = new HashMap<>();
```
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/launch/JavaLaunchCommandBuilder.java]: This change is unrelated, right? Maybe undo to keep the PR focused on the one change.

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java [hunk 4]
```diff
@@ -81,14 +81,14 @@ public YarnSubmissionHelper(final YarnConfiguration yarnConfiguration,
 
     LOG.log(Level.FINE, "Requesting Application ID from YARN.");
     this.yarnClientApplication = this.yarnClient.createApplication();
-    this.applicationResponse = yarnClientApplication.getNewApplicationResponse();
-    this.applicationSubmissionContext = yarnClientApplication.getApplicationSubmissionContext();
-    this.applicationId = applicationSubmissionContext.getApplicationId();
+    this.applicationResponse = this.yarnClientApplication.getNewApplicationResponse();
```
Comment (markusweimer) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java]: All these changes are unrelated to the change at hand, right?

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/uploader/JobFolder.java [hunk 5]
```diff
@@ -57,53 +57,57 @@
...
-   * Uploads the given file to the DFS.
-   *
-   * @param localFile
+   * Upload given file to the DFS.
+   * @param localFile File on local FS to upload to the (remote) DFS.
    * @return the Path representing the file on the DFS.
-   * @throws IOException
+   * @throws IOException if unable to upload the file or local file cannot be read.
    */
   public Path upload(final File localFile) throws IOException {
```
Comment (markusweimer) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/uploader/JobFolder.java]: All changes in this file are orthogonal to the goals of the PR, right?

**Comment by motus** [comment]
@markusweimer I've split the readability and logging changes into a separate PR #1180
